File: Python/Problems/Problems/problem3312.py
import heapq
import math
import queue
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter, deque

from .problem import Problem

class Problem3312(Problem):
    def test(self) -> bool:
        nums = [2, 3, 4]
        queries = [0, 2, 2]

        expected = [1, 2, 2]

        result = self.gcdValues(nums, queries)

        print1DArray(result)

        return expected == result

    def gcdValues(self, nums: List[int], queries: List[int]) -> List[int]:
        n = len(nums)
        mx = 0
        for i in range(n):
            mx = max(mx, nums[i])
        freqs = [0] * (mx + 1)
        for i in range(n):
            freqs[nums[i]] += 1
        gcds = [0] * (mx + 1)
        for i in range(mx, 0, -1):
            pairs = 0
            bigger = 0
            for j in range(i, mx + 1, i):
                pairs += freqs[j]
                bigger += gcds[j]
            gcds[i] = pairs * (pairs - 1) / 2 - bigger
        for i in range(1, mx + 1):
            gcds[i] += gcds[i - 1]
        m = len(queries)
        result = []

        for i in range(m):
            l = 0
            r = mx
            val = queries[i] + 1
            while l < r:
                mid = (l + r) // 2
                if gcds[mid] < val:
                    l = mid + 1
                else:
                    r = mid

            result.append(l)

        return result

File: C++/Problems/Problems/problem_3312.cpp
#include "problem_3312.h"

bool problem_3312::test() {
    std::vector nums { 2, 3, 4 };
    std::vector<long long> queries  { 0, 2, 2 };

    std::vector expected { 1, 2, 2 };

    auto result = gcdValues(nums, queries);

    print1DVector(result);

    return expected == result;
}

std::vector<int> problem_3312::gcdValues(std::vector<int> &nums, std::vector<long long> &queries) {
    int n = nums.size();
    int mx = 0;
    for (int i = 0; i < n; ++i)
        mx = std::max(mx, nums[i]);
    std::vector freqs(mx + 1, 0);
    for (int i = 0; i < n; ++i)
        freqs[nums[i]]++;
    std::vector<int> gcds(mx + 1);
    for (int i = mx; i > 0; --i)
    {
        long pairs = 0;
        long bigger = 0;
        for (int j = i; j <= mx; j += i)
        {
            pairs += freqs[j];
            bigger += gcds[j];
        }
        gcds[i] = pairs * (pairs - 1) / 2 - bigger;
    }
    for (int i = 1; i <= mx; ++i)
        gcds[i] += gcds[i - 1];
    int m = queries.size();
    std::vector<int> result(m);

    for (int i = 0; i < m; ++i)
    {
        int l = 0;
        int r = mx;
        int val = queries[i] + 1;
        while (l < r)
        {
            int mid = (l + r) / 2;
            if (gcds[mid] < val)
                l = mid + 1;
            else
                r = mid;
        }

        result[i] = l;
    }

    return result;
}
